libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,600 @@
+/*
+   GCS MAVLink functions related to FTP
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <AP_HAL/AP_HAL.h>
+
+#include "GCS.h"
+
+#include <AP_Filesystem/AP_Filesystem.h>
+
+#if HAVE_FILESYSTEM_SUPPORT
+
+extern const AP_HAL::HAL& hal;
+
+struct GCS_MAVLINK::ftp_state GCS_MAVLINK::ftp;
+
+bool GCS_MAVLINK::ftp_init(void) {
+    // we can simply check if we allocated everything we need
+    if (ftp.requests != nullptr) {
+        return true;
+    }
+
+    ftp.requests = new ObjectBuffer<pending_ftp>(5);
+    if (ftp.requests == nullptr) {
+        goto failed;
+    }
+    ftp.replies = new ObjectBuffer<pending_ftp>(30);
+    if (ftp.replies == nullptr) {
+        goto failed;
+    }
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&GCS_MAVLINK::ftp_worker, void),
+                                      "FTP", 1024, AP_HAL::Scheduler::PRIORITY_IO, 0)) {
+        goto failed;
+    }
+
+    return true;
+
+failed:
+    delete ftp.requests;
+    ftp.requests = nullptr;
+    delete ftp.replies;
+    ftp.replies = nullptr;
+
+    return false;
+    
+}
+
+void GCS_MAVLINK::handle_file_transfer_protocol(const mavlink_message_t &msg) {
+    if(ftp_init()) {
+        mavlink_file_transfer_protocol_t packet;
+        mavlink_msg_file_transfer_protocol_decode(&msg, &packet);
+
+        struct pending_ftp request;
+
+        request.chan = chan;
+        request.seq_number = *(uint16_t *)packet.payload;
+
+        request.session = packet.payload[2];
+        request.opcode = static_cast<FTP_OP>(packet.payload[3]);
+        request.size = packet.payload[4];
+        request.req_opcode = static_cast<FTP_OP>(packet.payload[5]);
+        request.burst_complete = packet.payload[6];
+        request.offset = *(uint32_t *)(&packet.payload[8]);
+        request.sysid = msg.sysid;
+        request.compid = msg.compid;
+        memcpy(request.data, &packet.payload[12], sizeof(packet.payload) - 12);
+
+        if (!ftp.requests->push(request)) {
+            // dropping the message, no buffer space to queue it in
+            // we could NACK it, but that can lead to GCS confusion, so we're treating it like lost data
+        }
+    }
+}
+
+void GCS_MAVLINK::send_ftp_replies(void) {
+    if (ftp.replies != nullptr) {
+        for (uint8_t i = 0; i < 20; i++) {
+            struct pending_ftp reply;
```

Comment:
This is a *lot* of data on the stack here! 500-odd bytes?

Reply:
MAVLink gets a large stack for this reason. There doesn't appear to be a good way around this at the moment.